File: forge/src/main/java/org/bread_experts_group/breadlib/platform/ForgePlatformHelper.java
package org.bread_experts_group.breadlib.platform;

import net.minecraft.network.protocol.common.custom.CustomPacketPayload;
import net.minecraft.server.level.ServerLevel;
import net.minecraft.server.level.ServerPlayer;
import net.minecraft.world.level.ChunkPos;
import net.minecraftforge.api.distmarker.Dist;
import net.minecraftforge.fml.ModList;
import net.minecraftforge.fml.loading.FMLLoader;
import net.minecraftforge.fml.loading.FMLPaths;
import net.minecraftforge.network.PacketDistributor;
import org.bread_experts_group.breadlib.ForgeNetworking;

import java.nio.file.Path;

public class ForgePlatformHelper implements IPlatformHelper {
	@Override
	public String getPlatformName() {
		return "Forge";
	}

	@Override
	public Path getConfigDir() {
		return FMLPaths.CONFIGDIR.get();
	}

	@Override
	public boolean isModLoaded(String modId) {
		return ModList.get().isLoaded(modId);
	}

	@Override
	public EnvironmentKind getEnvironmentKind() {
		if (FMLLoader.isProduction()) {
			return EnvironmentKind.RELEASE;
		} else {
			return EnvironmentKind.DEVELOPMENT;
		}
	}

	@Override
	public ApplicationSide getSide() {
		if (FMLLoader.getDist() == Dist.CLIENT) {
			return ApplicationSide.CLIENT;
		} else {
			return ApplicationSide.SERVER;
		}
	}

	@Override
	public void sendToServer(CustomPacketPayload payload) {
		ForgeNetworking.checkChannelNotNull();
		ForgeNetworking.NETWORK_CHANNEL.send(payload, PacketDistributor.SERVER.noArg());
	}

	@Override
	public void sendToAllPlayers(CustomPacketPayload payload, ServerLevel level) {
		ForgeNetworking.checkChannelNotNull();
		ForgeNetworking.NETWORK_CHANNEL.send(payload, PacketDistributor.ALL.noArg());
	}

	@Override
	public void sendToPlayersTrackingChunk(CustomPacketPayload payload, ServerLevel level, ChunkPos pos) {
		ForgeNetworking.checkChannelNotNull();
		for (ServerPlayer player : level.getChunkSource().chunkMap.getPlayers(pos,false)) {
			ForgeNetworking.NETWORK_CHANNEL.send(payload, PacketDistributor.PLAYER.with(player));
		}
	}

	@Override
	public void sendToPlayersInDimension(CustomPacketPayload payload, ServerLevel level) {
		ForgeNetworking.checkChannelNotNull();
		ForgeNetworking.NETWORK_CHANNEL.send(payload, PacketDistributor.DIMENSION.with(level.dimension()));
	}
}

File: forge/src/main/kotlin/org/bread_experts_group/breadlib/platform/ForgePlatformHelper.kt
package org.bread_experts_group.breadlib.platform

import net.minecraftforge.api.distmarker.Dist
import net.minecraftforge.fml.ModList
import net.minecraftforge.fml.loading.FMLLoader
import net.minecraftforge.fml.loading.FMLPaths
import java.nio.file.Path

class ForgePlatformHelper : IPlatformHelper {
	override val platformName: String = "Forge"
	override val configDir: Path
		get() = FMLPaths.CONFIGDIR.get()
	override val gameDir: Path
		get() = FMLPaths.GAMEDIR.get()
	override val environmentKind: EnvironmentKind
		get() = if (FMLLoader.isProduction()) EnvironmentKind.RELEASE else EnvironmentKind.DEVELOPMENT
	override val side: ApplicationSide
		get() = if (FMLLoader.getDist() == Dist.CLIENT) ApplicationSide.CLIENT else ApplicationSide.SERVER

	override fun isModLoaded(modId: String): Boolean = ModList.get().isLoaded(modId)

	override fun getModInfo(modId: String): ModInfo {
		val container = ModList.get().getModContainerById(modId).get()
		val info = container.modInfo
		val dependencies = info.dependencies.map { it.modId }
		val path = info.owningFile.file.filePath
		val version = info.owningFile.versionString()

		return ModInfo(modId, info.description, version, dependencies, path)
	}
}